src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java

```diff
@@ -146,48 +202,68 @@ public T apply(@Nullable Tuple2<I, Iterable<L>> input) {
-        List<PartitionLocatable<I>> indexedIntervals = new ArrayList<>();
-        int lastPartitionStartIndex = 0;
-        for (I interval : intervals) {
-            int[] partitionIndexes = overlapDetector.getOverlaps(interval).stream()
+        Broadcast<OverlapDetector<PartitionLocatable<SimpleInterval>>> overlapDetectorBroadcast = ctx.broadcast(overlapDetector);
+        JavaRDD<PartitionLocatable<I>> indexedIntervals = intervals.map(interval -> {
+            int[] partitionIndexes = overlapDetectorBroadcast.getValue().getOverlaps(interval).stream()
                     .mapToInt(PartitionLocatable::getPartitionIndex).toArray();
             if (partitionIndexes.length == 0) {
-                // interval does not overlap any partition - add it to the last one
-                indexedIntervals.add(new PartitionLocatable<>(lastPartitionStartIndex, interval));
-                continue;
+                // interval does not overlap any partition - add it to the one after the interval start
+                int i = Collections.binarySearch(firstLocatables, new SimpleInterval(interval), (o1, o2) -> IntervalUtils.compareLocatables(o1, o2, sequenceDictionary));
+                if (i >= 0) {
+                    throw new IllegalStateException(); // TODO: no overlaps, yet start of interval matches a partition read extent start
+                }
+                int insertionPoint = -i - 1;
+                if (insertionPoint == firstLocatables.size()) {
+                    insertionPoint = firstLocatables.size() - 1;
+                }
+                return new PartitionLocatable<>(insertionPoint, interval);
             }
             Arrays.sort(partitionIndexes);
             int startIndex = partitionIndexes[0];
             int endIndex = partitionIndexes[partitionIndexes.length - 1];
-            indexedIntervals.add(new PartitionLocatable<I>(startIndex, interval));
+            return new PartitionLocatable<>(startIndex, endIndex, interval);
+        });
+
+        // Create an RDD of intervals with the same number of partitions as the locatables, and where each interval
+        // is in its start partition. Within each partition, intervals are sorted by IntervalUtils#compareLocatables.
+        JavaRDD<PartitionLocatable<I>> indexedIntervalsRepartitioned = indexedIntervals
+                .mapToPair(interval ->
+                        new Tuple2<>(interval, (Void) null))
+                .repartitionAndSortWithinPartitions(new PartitionLocatablePartitioner(locatables.getNumPartitions()), new PartitionLocatableComparator<I>(sequenceDictionary))
+                .keys();
+
+        indexedIntervalsRepartitioned.cache(); // cache since we need to do two calculations on the intervals
+
+        Map<Integer, Integer> maxEndPartitionIndexesMap = indexedIntervalsRepartitioned.mapToPair((PairFunction<PartitionLocatable<I>, Integer, Integer>) partitionLocatable ->
```

Comment:
I think this method generally should have some line comments in it to help readability

Reply:
Done